go-controller/pkg/ovn/ovn.go

```diff
@@ -131,6 +141,31 @@ type Controller struct {
 	// Supports multicast?
 	multicastSupport bool
 
+	// Interface used for programming OVN for egress IP, based on the mode it's running in.
+	modeEgressIP modeEgressIP
+
+	// Sync used for retrying EgressIP objects which were created before any node existed.
+	egressAssignmentRetry sync.Map
+
+	// Mutex used for syncing the egressIP namespace handlers
+	egressIPNamespaceHandlerMutex *sync.Mutex
+
+	// Cache used for keeping track of EgressIP namespace handlers
+	egressIPNamespaceHandlerCache map[string]factory.Handler
+
+	// Mutex used for syncing the egressIP pod handlers
+	egressIPPodHandlerMutex *sync.Mutex
+
+	// Cache used for keeping track of EgressIP pod handlers
+	egressIPPodHandlerCache map[string]factory.Handler
+
+	// A cache used for egress IP assignments containing data for all cluster nodes
+	// used for egress IP assignments
+	eIPAllocator map[string]*eNode
+
+	// A mutex for eIPAllocator
+	eIPAllocatorMutex *sync.Mutex
+
```

Comment:
it would be nice if, instead of having a whole bunch of egressIP-related fields in `Controller`, if they could all be split off into some separate object

Reply:
It would be nice if go had generics too. Which is the entire problem here: we need to distinguish shared vs. local on the master side for this. I can only do that by specifying the interface `modeEgressIP`, but I can't have a generic struct which contains all the fields I now have in `Controller`. I need something that acts as a container which is readily available from any object inheriting it, which without generics becomes `Controller`.